Update product launch stages article (#221)
apollographql/docs: src/content/basics/resources/product-launch-stages.mdx

### Preview

Product and feature **previews** are usually announced publicly and represent a near-complete version of a feature set. However, previews are _not_ officially supported with any SLA, and breaking API changes are possible between incremental versions of a preview.

**A preview might still contain bugs or undergo iteration.** You're encouraged to try previews in test environments to familiarize yourself with upcoming functionality before it reaches [general availability](#general-availability).

### General availability

A **generally available** product or feature has been deemed ready for use in a production environment and is officially supported by Apollo. Its documentation is considered sufficient to support widespread adoption.

## Open-source pre-releases

Apollo open-source projects might provide pre-releases for early access:

### Alpha / Beta

An **alpha** or **beta** is in volatile, active development. The version might not be feature-complete, and breaking API changes are possible between incremental versions of the alpha or beta. 

> These stages help Apollo gather feedback and issues from community members and customers who are enthusiastic early adopters.

Reviewer: ```suggestion
Pre-releases help Apollo gather feedback and issues from community members and customers who are enthusiastic early adopters.
```

### Release candidate (RC)

> A **release candidate** (RC) version is considered potentially viable for general availability. Minor bugs might still be present, and documentation for the release might be incomplete.

Reviewer: ```suggestion
A **release candidate** (RC) is considered potentially viable for general availability. Minor bugs might still be present, and documentation for the release might be incomplete.
```

You're encouraged to test out RC versions to help Apollo identify any remaining issues.

## Experimental features

> At Apollo, we believe in developing and iterating in the open. **Experimental features** are targeted at developers who want to adopt advanced, cutting-edge functionality. These features might not be complete, and breaking API changes are possible between individual versions.

Reviewer: ```suggestion
At Apollo, we believe in developing and iterating in the open. **Experimental features** are targeted at developers who want to adopt advanced, cutting-edge functionality. These features might not be complete, may have breaking API changes at any time, and may be removed from a product or new version of an open-source release without notice.
```

> Experimental features are in active development. They can be present in a product that's in _any_ of the stages defined above, and should be used cautiously. **Experimental features might be removed from a product at any time without notice.**

Reviewer: ```suggestion
```

we can probably shorten this to the paragraph above, to keep this as short as possible and because most folks know what an experimental feature is.

## Stages for discontinuing support

> When support for a [generally available](#general-availability) Apollo product or feature is ending, that release proceeds through two additional stages: **deprecated** and **end of life**.

Reviewer: ```suggestion
When support for a [generally available](#general-availability) Apollo product or feature is ending, it becomes **deprecated** and then **end of life**.
```

Reviewer: **is ending** seems fine to me since it's in GA, and then when support is ending: 
1) first the feature becomes deprecated and an EOL date is provided, during which time **it still IS supported**
2) when it reaches EOL, then support does end, so support has not ended until it reaches the EOL date.

Reviewer 2: Good point, these can overlap in terms of timelines.

### Deprecated

> A **deprecated** release is still supported by Apollo. It continues to receive security patches and updates to address major regressions. After a specified date, deprecated releases move to [end-of-life](#end-of-life), at which point the release is no longer supported.

Reviewer: ```suggestion
A **deprecated** product or feature is still supported by Apollo and continues to receive security patches and updates to address major regressions. When deprecated, an [end-of-life](#end-of-life) date will be provided after which the product or feature will no longer be supported.
```

Reviewer: think the current wording is good. regressions are a subset of issues focused on ensuring things keep working as opposed to issues which are more broad. when something becomes deprecated we want to limit the scope to regressions as no new development will be done on a deprecated thing.

**Avoid relying on deprecated releases whenever possible.** Consider upgrading to the latest generally available release to continue receiving new features and functionality.

### End-of-life

An **end-of-life** release is **no longer supported or maintained by Apollo**. These releases no longer receive security patches or remediation for regressions. There is no commitment from Apollo to release further versions and no guarantee that the most recent version will continue to work.

> **Do not run end-of-life releases in production.** Upgrade to the latest generally available release as soon as possible to ensure you continue receiving security patches, bug fixes, and new features. 

Reviewer: ```suggestion
**Do not run end-of-life open-source releases in production.** Upgrade to the latest generally available  release as soon as possible to ensure you continue receiving security patches, bug fixes, and new features. 
```

---
title: Launch stages for Apollo products and features
---

> ## Universal launch stages

Reviewer 2: This term universal isn't clear to me, universal in regards to what specifically?

Reviewer: Agree. **Product launch stages** would be more clear.

Side note: There is launch, steady state, and descent. For launch we have Preview and GA. The other things are not part of Product launch.

All new Apollo products, features, and open-source libraries use two primary launch stages: **preview** and **general availability**.

### Preview

> Product and feature **previews** are usually announced publicly and represent a near-complete version of the release. However, previews are _not_ officially supported with any SLA, and breaking API changes are possible between versions.

Reviewer 2: This seems to contradict itself.  It's "near complete" but it may also break?

> Product and feature **previews** are usually announced publicly and represent a near-complete version of the release. However, previews are _not_ officially supported with any SLA, and breaking API changes are possible between versions.

Reviewer: would be fine with either one, up to @StephenBarlow 🙏

**A preview might still contain bugs or undergo iteration.** You're encouraged to try out preview releases in test environments to familiarize yourself with upcoming functionality before it reaches [general availability](#general-availability).

### General availability

A **generally available** product or feature has been deemed ready for use in a production environment and is officially supported by Apollo. Its documentation is considered sufficient to support widespread adoption.

> ## Pre-release stages

Reviewer 2: Aren't these part of our alpha / beta cycles? Like Beta 2, Beta 3, etc?

I guess i'm a bit confused on why this is outside of the Alpha  Beta section and it's just a bit confusing to read that any one time we could have features in 4 - 5 different different states of release.

A particular product or feature might include one or more additional pre-release stages in its development lifecycle:

### Alpha / Beta

An **alpha** or **beta** release is in volatile, active development. The release might not be feature-complete, and breaking API changes are possible between individual versions. 

These stages help Apollo gather feedback and issues from community members and customers who are enthusiastic early adopters.

> ### Release candidate (RC)

Reviewer 2: I'd suggest this goes below GA, as an RC is the step for GA, at least in OSS.

Reviewer: would keep this here as it's an OSS release stage, not a product launch stage

A **release candidate** (RC) release is considered potentially viable for general availability. Minor bugs might still be present, and documentation for the release might be incomplete.

You're encouraged to test out RC releases to help Apollo identify any remaining issues.

## Stages for discontinuing support

When support for a [generally available](#general-availability) Apollo product or feature is ending, that release proceeds through two additional stages: **deprecated** and **end of life**.

### Deprecated

A **deprecated** release is still supported by Apollo. It continues to receive security patches and updates to address major regressions. After a specified date, deprecated releases move to [end-of-life](#end-of-life), at which point the release is no longer supported.

**Avoid relying on deprecated releases whenever possible.** Consider upgrading to the latest generally available release to continue receiving new features and functionality.

### End-of-life

> An **end-of-life** release is **no longer supported or maintained by Apollo**. These releases no longer receive security patches or remediation for regressions. There is no commitment from Apollo to release further versions and no guarantee that the most recent version will continue to work.

Reviewer: Regressions seems more broad here, in terms of any regression no matter how small, will not be remediated which is what we want for EOL -- so would be fine keeping the current "regression" language.

**Do not run end-of-life releases in production.** Upgrade to the latest generally available release as soon as possible to ensure you continue receiving security patches, bug fixes, and new features. 

> ## Experimental features

Reviewer 2: I'd suggest we move this above the Stages for discontinued support as Experimental is an active stage of development, and sequentially  it seems out of order below the sections discussing EOL.

Reviewer: Agree, good catch. 👍
